### PR TITLE
rpc: force TLS for both grpc endpoint and grpc gateway

### PR DESCRIPTION
**Pull request by aakselrod**
This PR forces the grpc endpoint and proxy to use TLS. In the absence of a keypair/certificate, a self-signed keypair/certificate is generated on `lnd` startup.

`lncli` validates against a certificate or CA file chosen by (in order of decreasing preference)
* the `tlscertpath` option on the command line,
* the `tlscertpath` option in the config file (can optionally be specified by the `config` option on the command line), or
* the default `tls.cert` file in the `lnd` home directory.

Only the highest priority option which is specified is checked by `lncli`; it cannot check against 3 separate certificates/CAs in a single run.

### CODE REVIEW

File: cmd/lncli/main.go [hunk 1]
```diff
@@ -25,12 +41,37 @@ func getClient(ctx *cli.Context) (lnrpc.LightningClient, func()) {
 	return lnrpc.NewLightningClient(conn), cleanUp
 }
 
+type config struct {
+	SSLCert string `long:"sslcert" description:"path to SSL certificate"`
```
Comment (PirosB3) [inline cmd/lncli/main.go]: To be more descriptive, shouldn't this be named `SSLCertPath`?

File: cmd/lncli/main.go [hunk 2]
```diff
@@ -88,3 +139,17 @@ func main() {
...
 	}
 }
+
+// cleanAndExpandPath expands environment variables and leading ~ in the
+// passed path, cleans the result, and returns it.
+func cleanAndExpandPath(path string) string {
+	// Expand initial ~ to OS specific home directory.
+	if strings.HasPrefix(path, "~") {
+		homeDir := filepath.Dir(lndHomeDir)
+		path = strings.Replace(path, "~", homeDir, 1)
```
Comment (PirosB3) [inline cmd/lncli/main.go]: An edge case: if `path` contains `~` as part of file name or dir name (ex. `~/hello~world/ssl.cert`) this will replace more than once, leading to a bug

File: lnd.go [hunk 3]
```diff
@@ -278,3 +297,37 @@ func main() {
...
+	}
+	return true
+}
+
+// genCertPair generates a key/cert pair to the paths provided.
+func genCertPair(certFile, keyFile string) error {
+	rpcsLog.Infof("Generating TLS certificates...")
+
+	org := "lnd autogenerated cert"
+	validUntil := time.Now().Add(10 * 365 * 24 * time.Hour)
```
Comment (PirosB3) [inline lnd.go]: should `10 * 365 * 24 * time.Hour` be assigned to a constant? say `AUTOGENERATED_CERT_SECONDS`?

File: lnd.go [hunk 4]
```diff
@@ -278,3 +297,37 @@ func main() {
...
+	if err != nil {
+		return err
+	}
+
+	// Write cert and key files.
+	if err = ioutil.WriteFile(certFile, cert, 0644); err != nil {
+		return err
+	}
+	if err = ioutil.WriteFile(keyFile, key, 0600); err != nil {
+		os.Remove(certFile)
```
Comment (PirosB3) [inline lnd.go]: why are you deleting the certificate?

File: networktest.go [hunk 5]
```diff
@@ -243,8 +247,16 @@ func (l *lightningNode) Start(lndError chan error) error {
 		return err
 	}
 
+	// Wait until SSL certificate is created before using it
+	for !fileExists(l.cfg.SSLCert) {
+		time.Sleep(100 * time.Millisecond)
```
Comment (PirosB3) [inline networktest.go]: this could ideally lock forever, right? have you thought of panicking after a max number of retries?

File: cmd/lncli/main.go [hunk 6]
```diff
@@ -3,11 +3,27 @@ package main
...
+	"google.golang.org/grpc/credentials"
+)
+
+const (
+	defaultConfigFilename  = "lnd.conf"
+	defaultSSLCertFilename = "ssl.cert"
+)
+
+var (
+	lndHomeDir        = btcutil.AppDataDir("lnd", false)
```
Comment (halseth) [inline cmd/lncli/main.go]: This already exists in `config.go`, can it be reused?

Comment (halseth) [inline cmd/lncli/main.go]: Ah, maybe you have no access to it since it is in the lnd/main package... Still, it would be nice to have it located in a place where it could be accessed both from lnd/main and from lncli.

Comment (Roasbeef) [inline cmd/lncli/main.go]: Yeah it can't be used as is currently. However, when we restructure the packages within `lnd` to accommodate your `gomobile` work @halseth we should be able to re-use the definitions.

File: cmd/lncli/main.go (hunk 1 above)
Comment (Roasbeef) [inline cmd/lncli/main.go]: Ditto on this comment.

File: cmd/lncli/main.go [hunk 7]
```diff
@@ -25,12 +41,37 @@ func getClient(ctx *cli.Context) (lnrpc.LightningClient, func()) {
...
 func getClientConn(ctx *cli.Context) *grpc.ClientConn {
 	// TODO(roasbeef): macaroon based auth
 	// * http://www.grpc.io/docs/guides/auth.html
 	// * http://research.google.com/pubs/pub41892.html
 	// * https://github.com/go-macaroon/macaroon
-	opts := []grpc.DialOption{grpc.WithInsecure()}
+	cfg := config{
+		SSLCert: defaultSSLCert,
+	}
+	// We want only the SSL certificate information from the configuration
```
Comment (Roasbeef) [inline cmd/lncli/main.go]: Formatting nit: a new line should be inserted above this line.

File: cmd/lncli/main.go [hunk 8]
```diff
@@ -25,12 +41,37 @@ func getClient(ctx *cli.Context) (lnrpc.LightningClient, func()) {
...
+	// file at this time, so ignore anything else. We can always add fields
+	// as we need them. When specifying a file on the `lncli` command line,
+	// this should work with just a trusted CA cert assuming the server's
+	// cert file contains the entire chain from the CA to the server's cert.
+	parser := flags.NewParser(&cfg, flags.IgnoreUnknown)
+	iniParser := flags.NewIniParser(parser)
```
Comment (Roasbeef) [inline cmd/lncli/main.go]: Why is this necessary? The `sslcert` flag will be parsed by the command line utility package just as all the other global level arguments.

Comment (Roasbeef) [inline cmd/lncli/main.go]: Ah never mind, I see now that it attempts to read `lnd's` configuration in order to obtain the path for the TLS cert.

File: cmd/lncli/main.go (hunk 1 above)
Comment (Roasbeef) [inline cmd/lncli/main.go]: An un-related comment: this should read `TLSCertPath`. `SSL` is insecure and shouldn't be used by anyone.

File: networktest.go [hunk 9]
```diff
@@ -308,13 +320,11 @@ func (l *lightningNode) Stop() error {
...
 	}
 
-	if runtime.GOOS == "windows" {
-		if err := l.cmd.Process.Signal(os.Kill); err != nil {
-			return err
-		}
-	} else if err := l.cmd.Process.Signal(os.Interrupt); err != nil {
-		return err
-	}
+	// Don't watch for error because sometimes the RPC connection gets
```
Comment (Roasbeef) [inline networktest.go]: This portion of the diff should be omitted once this PR is rebased to master.

File: config.go [hunk 10]
```diff
@@ -78,6 +82,8 @@ type config struct {
 
 	ConfigFile string `long:"C" long:"configfile" description:"Path to configuration file"`
 	DataDir    string `short:"b" long:"datadir" description:"The directory to store lnd's data within"`
+	SSLCert    string `long:"sslcert" description:"Path to SSL certificate for lnd's RPC and REST services"`
```
Comment (Roasbeef) [inline config.go]: Similar comments, here and the default variables above: ssl -> tls

File: lnd.go (hunk 4 above)
Comment (Roasbeef) [inline lnd.go]: Well the operation itself didn't complete cleanly. Therefore, in order to avoid any issues with a partially generated TLS state, I think deleting the cert file is the prudent thing to do.

File: cmd/lncli/main.go (hunk 2 above)
Comment (aakselrod) [inline cmd/lncli/main.go]: This is incorrect, see https://golang.org/pkg/strings/#Replace

The `1` parameter means that only the first instance of `~` gets replaced.

File: lnd.go (hunk 4 above)
Comment (aakselrod) [inline lnd.go]: This is correct; the certificate that's being deleted is the one that was just written in the case that the key for it can't also be written.

File: cmd/lncli/main.go (hunk 6 above)
Comment (aakselrod) [inline cmd/lncli/main.go]: Yep, this is correct.

**Comment by stevenroose** [comment]
Are you guys open to a `--notls` option?
